### PR TITLE
Read ESRI tilemaps in vtpk and arcgis tile services

### LINKED ISSUE #52874: ESRI .vtpk tiles are not rendered correctly when .vtpk is indexed
### What is the bug or the crash?

Hi,
I created .vtpk from sample .shp using ArcGIS Pro 2.9.1 using ArcGIS Pro 'Create Vector Tiles Package' GP tool.
The .vtpk was indexed. Index was created using ArcGIS Pro 'Create Vector Tiles Index' GP tool. Maximum vertex count have been set to 10000 vertexes and was based on ESRI WGS84_Geographic_Coordinate_System_V2.xml tiling scheme.
(.vtpk, orig.shp, index.shp and .xml attached). Then I dragged and dropped .vtpk into QGS Desktop 3.28.1. When I zoomed in into higher scales some of the tiles are not rendered. I guess the root of the problem lies in the fact that the QGIS Desktop map viewer searches tiles in current zoom but an indexed .vtpk fetches tiles from upper zooms (if vertex count in the current zoom is less than 10000), an action that QGIS Desktop map viewer failes to execute.
I tried to upgrade ArcGIS Pro version (3.0.2) and QGIS Desktop version (3.30.1) but the problem still occurs.
Thanks in advance,
Jossef
[IndexedVtpk.zip](https://github.com/qgis/QGIS/files/11308552/IndexedVtpk.zip)
image

### Steps to reproduce the issue

Above

### Versions

Above

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nyalldawson**
This allows us to correctly handle "indexed" vector tile sets, where higher zoom level tiles may be missing where a lower
zoom level tile is deemed to have sufficient detail for higher zoom levels.

Fixes https://github.com/qgis/QGIS/issues/52874

Funded by Landesamt für Vermessung und Geoinformation, Feldkirch, Austria

### CODE REVIEW

File: src/core/qgis.h [hunk 1]
```diff
@@ -3553,6 +3553,57 @@ class CORE_EXPORT Qgis
...
+    Q_DECLARE_FLAGS( VectorTileProviderFlags, VectorTileProviderFlag )
+    Q_FLAG( VectorTileProviderFlags )
+
+    /**
+     * Enumeration with capabilities that vector tile data providers might implement.
+     * \since QGIS 3.32
+     */
+    enum class VectorTileProviderCapability : int
+    {
+      ReadLayerMetadata = 1 << 1, //!< Provider can read layer metadata from data store. See QgsDataProvider::layerMetadata()
```
Comment (nirvn) [inline src/core/qgis.h:3581]: Was that added in 3.32 to begin with or do you need a SIP_MONKEYPATCH_SCOPEENUM_UNNEST?

File: src/core/vectortile/qgsvectortiledataprovider.h [hunk 2]
```diff
@@ -41,19 +41,6 @@ class CORE_EXPORT QgsVectorTileDataProvider : public QgsDataProvider
 
   public:
 
-    /**
-     * Enumeration with capabilities that vector tile data providers might implement.
-     * \since QGIS 3.32
```
Comment (nirvn) [inline src/core/vectortile/qgsvectortiledataprovider.h:46]: Ignore question above, I've got my answer :)

File: src/core/vectortile/qgsvectortilelayerrenderer.cpp [hunk 3]
```diff
@@ -26,7 +26,7 @@
 #include "qgsmapclippingutils.h"
 #include "qgsrendercontext.h"
 #include "qgsvectortiledataprovider.h"
-
+#include "qgstextrenderer.h"
 #include <QElapsedTimer>
```
Comment (nirvn) [inline src/core/vectortile/qgsvectortilelayerrenderer.cpp:30]:
```suggestion

#include <QElapsedTimer>
```

File: src/core/vectortile/qgsvectortilelayerrenderer.cpp [hunk 4]
```diff
@@ -278,9 +280,14 @@ void QgsVectorTileLayerRenderer::decodeAndDrawTile( const QgsVectorTileRawData &
     ctx.painter()->setPen( pen );
     ctx.painter()->setBrush( brush );
     ctx.painter()->drawPolygon( tile.tilePolygon() );
-#if 0
-    ctx.painter()->setBrush( QBrush( QColor( 255, 0, 0 ) ) );
-    ctx.painter()->drawText( tile.tilePolygon().boundingRect().center(), tile.id().toString() );
+#if 1
```
Comment (nirvn) [inline src/core/vectortile/qgsvectortilelayerrenderer.cpp:283]:
```suggestion
#if 0
```

File: src/core/vectortile/qgsvtpkvectortiledataprovider.cpp [hunk 5]
```diff
@@ -208,11 +213,32 @@ QList<QgsVectorTileRawData> QgsVtpkVectorTileDataProvider::readTiles( const QgsT
...
+
+  return metadata;
+}
+
 QByteArray QgsVtpkVectorTileDataProvider::loadFromVtpk( QgsVtpkTiles &vtpkTileReader, const QgsTileXYZ &id, QgsFeedback * )
 {
   const QByteArray tileData = vtpkTileReader.tileData( id.zoomLevel(), id.column(), id.row() );
   if ( tileData.isEmpty() )
   {
+    // TODO -- I think here ESRI software will detect a zero size tile and automatically fallback to lower zoom level tiles
```
Comment (nirvn) [inline src/core/vectortile/qgsvtpkvectortiledataprovider.cpp:238]: What's your plan here? Follow up PR?

File: src/core/vectortile/qgsvectortilelayerrenderer.cpp (hunk 4 above)
Comment (nyalldawson) [reply src/core/vectortile/qgsvectortilelayerrenderer.cpp:283]: I'd rather leave this -- it's only shown when the tile boundaries are visible, and that option is for debugging only and isn't exposed anywhere in the application.

File: src/core/vectortile/qgsvtpkvectortiledataprovider.cpp (hunk 5 above)
Comment (nyalldawson) [reply src/core/vectortile/qgsvtpkvectortiledataprovider.cpp:238]: Maybe address during freeze, if there's no higher priority bugs...

**Pull request merged**